spec/chrome_integration/gtalk.spec.js

```diff
@@ -0,0 +1,237 @@
+var CLIENT_ID =
+    '746567772449-jkm5q5hjqtpq5m9htg9kn0os8qphra4d.apps.googleusercontent.com';
+
+// Next access token to be returned by OAuthView.launchAuthFlow
+var pendingAccessToken = null;
+
+// Replace default OAuthView with a view that sets the access_token to
+// pendingAccessToken
+var OAuthView = function() {};
+OAuthView.prototype.initiateOAuth = function(redirectURIs, continuation) {
+  continuation({redirect: REDIRECT_URL, state: ''});
+  return true;
+};
+OAuthView.prototype.launchAuthFlow = function(authUrl, stateObj, continuation) {
+  if (!pendingAccessToken) {
+    continuation(undefined, 'No access token found');
+  } else {
+    continuation(REDIRECT_URL + '?access_token=' + pendingAccessToken);
+    pendingAccessToken = null;
+  }
+};
+
+// Returns a Promise that fulfills with an access token.
+function getAccessToken(refreshToken) {
+  return new Promise(function(fulfill, resolve) {
+    var data = 'refresh_token=' + refreshToken +
+        '&client_id=' + CLIENT_ID +
+        '&client_secret=h_hfPI4jvs9fgOgPweSBKnMu' +
+        '&grant_type=refresh_token';
+    ajax('POST', 'https://www.googleapis.com/oauth2/v3/token', data)
+        .then(function(resp) {
+          fulfill(resp.access_token);
+        });
+  });
+}
+
+// Returns a promise that fulfills when socialClient has been logged in
+// using the loginEmail address.  loginEmail must have an associated
+// refresh token.
+function loginAs(socialClient, loginEmail) {
+  var refreshToken = REFRESH_TOKENS[loginEmail.toLowerCase()];
+  if (!refreshToken) {
+    return Promise.reject('No refresh token found for ' + loginEmail);
+  } else if (pendingAccessToken) {
+    return Promise.reject('loginAs cannot be called concurrently.' +
+        '  Please wait until previous login completes');
+  }
+  return getAccessToken(refreshToken).then(function(accessToken) {
+    // Set the accessToken to pendingAccessToken so that it can be used by
+    // our OAuthView.
+    pendingAccessToken = accessToken;
+    return socialClient.login({
+        agent: 'integration',
+        version: '0.1',
+        url: '',
+        interactive: false,
+        rememberLogin: false
+      });
+  });
+}
+
+// Sets up an onClientState listener and fulfills the returned Promise
+// when the given userId appears as ONLINE.
+function waitForUser(socialClient, userId) {
+  return new Promise(function(fulfill, reject) {
+    socialClient.on('onClientState', function(clientState) {
+      if (clientState.userId == userId &&
+          clientState.status == 'ONLINE') {
+        fulfill(clientState);
+      }
+    });
+  });
+}
+
+describe('GTalk', function() {
+  var socialInterface;
+
+  // Message to be sent between peers.  If a unique message is not used,
+  // messages from one persons test might interfere with another person who
+  // is running the same tests at the same time.
+  var uniqueMsg = Math.random().toString();
+
+  function logoutThenDone(socialClientsArray, done) {
```

Comment:
Some of these helper functions are outside the describe. Can we put them all in 1 place?
Maybe a var Helper = {waitForUser: ..., logoutThenDone: ...} etc?